client/client.go

```diff
@@ -2,24 +2,90 @@ package client
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"net/http"
+	"time"
 
 	"github.com/deepmap/oapi-codegen/pkg/securityprovider"
 	"github.com/hashicorp/go-cleanhttp"
 	"github.com/hashicorp/go-retryablehttp"
 	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
 )
 
+// last retry will wait 64 seconds (2^6)
+// will also listen to 'Retry-After' header
+
+const (
+	rpm              = 600
+	bucketsPerMinute = 60
+	maxRetries       = 6
+	minRetryWait     = 1 * time.Second
+	maxRetryWait     = 65 * time.Second
+)
+
+type RateLimitedClient struct {
+	client      *http.Client
+	rateLimiter *rate.Limiter
+}
+
+func (c *RateLimitedClient) Do(req *http.Request) (*http.Response, error) {
+	if c.rateLimiter == nil {
+		return nil, errors.New("rate limiter not set")
+	}
+	ctx := req.Context()
+	err := c.rateLimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func attentiveBackoff(min, max time.Duration, attemptNum int, resp *http.Response) time.Duration {
+	// Retry for rate limits and server errors.
+	if resp != nil && resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode >= 500 {
+		// Check for a 'Retry-After' header.
+		retryAfter := resp.Header.Get("Retry-After")
+		if retryAfter != "" {
+			if retryAfterSeconds, err := time.ParseDuration(retryAfter + "s"); err == nil {
+				return retryAfterSeconds
+			}
+		}
+
+		// If not check if there is a response in the body
+		var data map[string]interface{}
+		json.NewDecoder(resp.Body).Decode(&data)
+		defer resp.Body.Close()
+		retryAfter, ok := data["Retry-After"].(string)
+		if ok && retryAfter != "" {
+			if retryAfterSeconds, err := time.ParseDuration(retryAfter + "s"); err == nil {
+				return retryAfterSeconds
+			}
+		}
+
+		// otherwise use the default backoff
+	}
+	return retryablehttp.DefaultBackoff(min, max, attemptNum, resp)
+}
+
 func New(ctx context.Context, apiKey, apiEndpoint, version string, opts ...ClientOption) (*ClientWithResponses, error) {
 	bearerTokenProvider, bearerTokenProviderErr := securityprovider.NewSecurityProviderBearerToken(apiKey)
 	if bearerTokenProviderErr != nil {
 		return nil, bearerTokenProviderErr
 	}
 
 	retryClient := retryablehttp.NewClient()
-	retryClient.RetryMax = 3
+
+	retryClient.RetryMax = maxRetries
+	retryClient.RetryWaitMin = minRetryWait
+	retryClient.RetryWaitMax = maxRetryWait
+	retryClient.Backoff = attentiveBackoff
```

Comment:
so we were using this retry client already and were erroring on the 502s.

I assume we've concluded that we didn't retry for long enough, but just to check - do we need to enable retry on 5xx errors ?

Reply:
turns out we were configuring a client and throwing it in the bin. This bug has now been fixed.